File: src/main/java/algo/slidingWindow/MinimumSizeSubarraySum_209.java
package algo.slidingWindow;

public class MinimumSizeSubarraySum_209 {
  public static int minSubArrayLen2(int target, int[] nums) {
    int i = 0, j = 0;
    int len = nums.length;
    int size = len + 1;
    int sum = nums[0];

    while (i <= j && j < len - 1) {

      // enlarge window size to the right
      while (sum < target && j < len - 1) {
        sum += nums[++j];
      }
      size = Math.min(size, j - i + 1);

      // shrink window from left
      while (sum >= target && i <= j) {
        sum -= nums[i++];
      }
      size = Math.min(size, j - i + 1);
    }

    return size <= len ? size : 0;
  }

  public static int minSubArrayLen(int target, int[] nums) {
    int left = 0, right = 0;
    var sum = 0;
    var minLen = Integer.MAX_VALUE;
    while (right < nums.length) {
      sum += nums[right];
      if (sum >= target) {
        while (left < right && sum - nums[left] >= target) {
          sum -= nums[left];
          left++;
        }
        minLen = Math.min(minLen, right - left + 1);
      }
      right++;
    }
    return minLen == Integer.MAX_VALUE ? 0 : minLen;
  }

}

File: src/main/kotlin/algo/slidingWindow/MinimumSizeSubarraySum_209.kt
package algo.slidingWindow

class KtMinimumSizeSubarraySum_209 {
    fun minSubArrayLen(target: Int, nums: IntArray): Int {
        var left = 0; var right = 0
        var sum = 0
        var minLen = Int.MAX_VALUE
        while (right < nums.size) {
            sum += nums[right]
            if (sum >= target) {
                while (left < right && sum - nums[left] >= target) {
                    sum -= nums[left]
                    left++
                }
                minLen = minLen.coerceAtMost(right - left + 1)
            }
            right++
        }
        return if (minLen == Int.MAX_VALUE) 0 else minLen;
    }
}

fun main() {
    val subarray = KtMinimumSizeSubarraySum_209()
//    println(subarray.minSubArrayLen(7, intArrayOf(2,3,1,2,4,3)))
//    println(subarray.minSubArrayLen(4, intArrayOf(1,4,4)))
    println(subarray.minSubArrayLen(11, intArrayOf(1,2,3,4,5)))
}